Week 3 - Submission 1
A, B,  C and D are matrices given below:
        130  120  105                6  7               5  2  7                245
A =      4      3      1        B =  4  5         C = 3  5  1         D =     6
            2      5      1                8  9               7  2  1                    7
Using Numpy compute:
A+ C
A- C
AC
CD

Solve the matrix equation (linear system) for x:
A x = D
This system written out in full looks like:
130x +120y +105z = 245
4x +3y + z = 6
2x +5y +z = 7
and solves the following problem:
I'm going on a diet and have decided I need to eat exactly 245 calories, 6 grams of protein and 7 grams of fat.
Unfortunately, I open my cupboard to see that all I have is three boxes of cereal: Cheerios, Cinnamon Toast Crunch, and Rice Krispies. The nutritional information per serving is as follows:
Cereal                                   Calories     Protein     Fat
Cheerios                                     120          4           2
Cinnamon Toast Crunch             130          3           5
Rice Krispies                               105          1           2

First I denote c as the number of servings of Cheerios, t as the number of servings of Cinnamon Toast Crunch and 
r as the number of servings of Rice Krispies then the three equations above have to be true simultaneously.

In [1]:
# import libraries that are required
import numpy as np
import sys
import matplotlib.pyplot as plt
 #Setting up arrays for the problems  
A = np.array([[130,120,105],
              [4,3,1],
              [2,5,1]], dtype=float)
B = np.array([[6,7],
              [4,5],
              [8,9]], dtype=float)
C = np.array([[5,2,7],
              [3,5,1],
              [7,2,1]], dtype=float)
D = np.array([[245],
               [6],
               [7]], dtype=float)

In [2]:
A + C #matrix addition

array([[135., 122., 112.],
       [  7.,   8.,   2.],
       [  9.,   7.,   2.]])

In [3]:
A - C #matrix substition 

array([[125., 118.,  98.],
       [  1.,  -2.,   0.],
       [ -5.,   3.,   0.]])

In [4]:
np.matmul(A,C) #True Matrix Multiplication

array([[1745., 1070., 1135.],
       [  36.,   25.,   32.],
       [  32.,   31.,   20.]])

In [5]:
np.matmul(C,D) #True Matrix Multiplication

array([[1286.],
       [ 772.],
       [1734.]])

In [6]:
x = np.linalg.solve(A,D) #solving the system of equations (example problem)
x

array([[0.59278351],
       [1.09278351],
       [0.35051546]])

CEREAL

In [7]:
#Cheerios (3x1)
#Calories Protein Fat
c = np.array([[120],
              [4],
              [2]], dtype=float)

#Cinnamon Toast Crunch (3x1)
#Calories Protein Fat
t = np.array([[130],
              [3],
              [5]], dtype=float)

#Rice Krispies (3x1)
#Calories Protein Fat
r = np.array([[105],
              [1],
              [2]], dtype=float)

#All cereals 
ctr = np.array([[120,130,105],
              [4,3,1],
              [2,5,2]] , dtype=float)

#Diet (3x1)
d = np.array([[245],
             [6],
             [7]], dtype=float)

To find the x ,
A x = B
(A^-1) * A * x = B * (A^-1)
I * x = B * (A^-1)
x = (A^-1)(B)

In [8]:
AInverse = np.linalg.inv(ctr) # found the inverse of A
AInverse

array([[ 0.00123457,  0.32716049, -0.22839506],
       [-0.00740741,  0.03703704,  0.37037037],
       [ 0.01728395, -0.41975309, -0.19753086]])

In [9]:
v = np.matmul(AInverse,d) #multiply A inverse and d (diet).
v

array([[0.66666667],
       [1.        ],
       [0.33333333]])

So, to meet the diet amount, you would need to eat the following:
Cheerios: .66667 (2/3) - Serving
Cinnamon Toast Crunch: 1 - Serving
Rice Krispies .33333 (1/3) - Serving

In [11]:
np.matmul(A,v)

array([[241.66666667],
       [  6.        ],
       [  6.66666667]])

This is done using x = np.linalg.solve

In [12]:
x = np.linalg.solve(ctr,d)
x

array([[0.66666667],
       [1.        ],
       [0.33333333]])